### ISSUE TITLE
backupdir seems broken for symlink

### ISSUE THREAD

**Issue by jreybert** [neovim/neovim#18659]
### Neovim version (nvim -v)

NVIM v0.7.0

### Vim (not Nvim) behaves the same?

no, vim 8.0 1-1763

### Operating system/version

Rocky 8.5

### Terminal name/version

kitty 0.23.1

### $TERM environment variable

xterm-256color

### Installation

system package manager

### How to reproduce the issue

```vim
set backup
set backupdir=.backup/,~/.vim-backup/backup//,/tmp//
```

```console
$ cd tmp/; mkdir -p a b; touch a/foo; ln -fs ../a/foo b/foo
$ nvim a/foo # do some edit, write, edit, write
$ ls ~/.vim-backup/backup/
%home%jerome%tmp%a%foo~
$ nvim b/foo # do some edit, write
```

### Expected behavior

Write succeeds

Not sure about the behavior, but I would like something like

```console
$ ls ~/.vim-backup/backup/
%home%jerome%tmp%a%foo~
%home%jerome%tmp%b%foo~
```

vim 8.0 behavior is not an example, as it does not support complete path backup name with %

### Actual behavior

When I write in neovim, I get an error:

```
:w
E506: Can't write to backup file (add ! to override)
```

**Comment by jreybert** [comment]
And another issue, which seems related, and maybe even more serious.

If I set the option `set backupcopy=auto,breaksymlink`, it breaks the symlink of the original file, not of the backupfile.
I am not sure to read the documentation correctly, but it seems really agressive to transform a symlink to a proper file. It seems especially not related with backup strategy.

Example: initially, b/foo is a symlink of a/foo.
```console
$ ls -l a b
a:
total 4
-rw-rw-r--. 1 jerome jerome 27 May 20 15:45 foo

b:
total 0
lrwxrwxrwx. 1 jerome jerome 8 May 20 15:45 foo -> ../a/foo
```

If I configure neovim with `set backupcopy=auto,breaksymlink`:

```console
$ nvim b/foo # edit, write
$ ls -l a b
a:
total 4
-rw-rw-r--. 1 jerome jerome 27 May 20 15:45 foo

b:
total 4
-rw-rw-r--. 1 jerome jerome 27 May 20 15:48 foo
```

Is this really the expected behavior of breaksymlink?

**Comment by justinmk** [comment]
Can you review the previous closed issues and say if this is a regression or something different? https://github.com/neovim/neovim/issues?q=is%3Aissue+backupdir+symlink+is%3Aclosed

**Comment by jreybert** [comment]
TL;DR:
- https://github.com/neovim/neovim/issues/11349
  - this is it! but it states it is fixed in master since June 2021.
  - if I remove the first non existing directory in .vim.init, it works `set backupdir=~/.vim-backup/backup//,/tmp//
`

I tried with vim 8.2.2434, and it behaves like it seems natural to me:

```console
$ cd tmp/; mkdir -p a b; touch a/foo; ln -fs ../a/foo b/foo
$ mkdir -p ~/.vim-backup/backup
$ rm ~/.vim-backup/backup/*
$ cat .vim.init
set backup
set backupdir=.backup/,~/.vim-backup/backup//,/tmp//
$ vim a/foo
edit save edit save
$ ll ~/.vim-backup/backup/
total 8
-rw-r--r-- 1 jerome jerome 3 21 mai   21:11 %home%jerome%tmp%a%foo~
$ vim b/foo
edit save edit save
$ ll ~/.vim-backup/backup/
total 8
-rw-r--r-- 1 jerome jerome 3 21 mai   21:11 %home%jerome%tmp%a%foo~
-rw-r--r-- 1 jerome jerome 6 21 mai   21:11 %home%jerome%tmp%b%foo~
```

**Comment by justinmk** [comment]
## Status

1. rebase https://github.com/neovim/neovim/pull/11417 and see if CI passes now?
2. see if that PR fixes this issue

**Comment by jreybert** [comment]
Rebase done in #18704

For the CI, I let you judge. Lot of debug and temporary commits, not all the 6 commits should be merged.

The good news is that it seems to fix the issue. I made a little script horror to make sure I am testing the right thing:

```bash
#!/usr/bin/env bash

set -eu

EDITOR=$1

$EDITOR --version

TEST_DIR=test_nvim_backupdir
BACKUP_DIR=~/.test_nvim_backupdir//
INIT_FILE=.vim.init

rm -rf $TEST_DIR
mkdir $TEST_DIR
cd $TEST_DIR

mkdir -p a b
touch a/foo
ln -fs ../a/foo b/foo

rm -rf $BACKUP_DIR
mkdir -p $BACKUP_DIR

cat > .vim.init << EOF
set backup
set backupdir=.donotexistdir,$BACKUP_DIR,/tmp//
EOF

$EDITOR a/foo -u $INIT_FILE -c ":silent put ='edit 1'" -c ":w" -c ":silent put ='edit 2'" -c ":wq"
ls -l $BACKUP_DIR
$EDITOR b/foo -u $INIT_FILE -c ":silent put ='edit 1'" -c ":w" -c ":silent put ='edit 2'" -c ":wq" -c ":wq!"
ls -l $BACKUP_DIR
ls -l $BACKUP_DIR/*%a%foo~
ls -l $BACKUP_DIR/*%b%foo~ || (echo "ERROR b.foo has been backed up"; exit 1)
```
```console
jerome@vostro:~/tmp$ ./test_nvim.sh nvim
NVIM v0.8.0-dev+291-gb083dd196
Build type: Debug
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -Wimplicit-fallthrough -Wsuggest-attribute=pure -Wsuggest-attribute=const -Wsuggest-attribute=malloc -Wsuggest-attribute=cold -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=1 -DEXITFREE -I/home/jerome/src/neovim/build/config -I/home/jerome/src/neovim/src -I/home/jerome/src/neovim/.deps/usr/include -I/usr/include -I/home/jerome/src/neovim/build/src/nvim/auto -I/home/jerome/src/neovim/build/include
Compiled by jerome@vostro

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/home/jerome/.local/share/nvim"

Run :checkhealth for more info
total 4
-rw-r--r-- 1 jerome jerome 8 May 22 22:01 %home%jerome%tmp%test_nvim_backupdir%a%foo~
total 8
-rw-r--r-- 1 jerome jerome  8 May 22 22:01 %home%jerome%tmp%test_nvim_backupdir%a%foo~
-rw-r--r-- 1 jerome jerome 22 May 22 22:01 %home%jerome%tmp%test_nvim_backupdir%b%foo~
-rw-r--r-- 1 jerome jerome 8 May 22 22:01 /home/jerome/.test_nvim_backupdir///%home%jerome%tmp%test_nvim_backupdir%a%foo~
-rw-r--r-- 1 jerome jerome 22 May 22 22:01 /home/jerome/.test_nvim_backupdir///%home%jerome%tmp%test_nvim_backupdir%b%foo~
jerome@vostro:~/tmp$ ./test_nvim.sh ~/tmp/nvim
NVIM v0.7.0
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/gcc-11 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -Wimplicit-fallthrough -Wsuggest-attribute=pure -Wsuggest-attribute=const -Wsuggest-attribute=malloc -Wsuggest-attribute=cold -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/runner/work/neovim/neovim/build/config -I/home/runner/work/neovim/neovim/src -I/home/runner/work/neovim/neovim/.deps/usr/include -I/usr/include -I/home/runner/work/neovim/neovim/build/src/nvim/auto -I/home/runner/work/neovim/neovim/build/include
Compiled by runner@fv-az316-460

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "
/home/runner/work/neovim/neovim/build/nvim.AppDir/usr/share/nvim"

Run :checkhealth for more info
total 4
-rw-r--r-- 1 jerome jerome 8 May 22 22:02 %home%jerome%tmp%test_nvim_backupdir%a%foo~
total 4
-rw-r--r-- 1 jerome jerome 8 May 22 22:02 %home%jerome%tmp%test_nvim_backupdir%a%foo~
-rw-r--r-- 1 jerome jerome 8 May 22 22:02 /home/jerome/.test_nvim_backupdir///%home%jerome%tmp%test_nvim_backupdir%a%foo~
ls: cannot access '/home/jerome/.test_nvim_backupdir///*%b%foo~': No such file or directory
ERROR b.foo has been backed up
```
